Userland/Libraries/LibDSP/Processor.h

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright (c) 2021, kleines Filmröllchen <malu.bertsch@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include "Music.h"
+#include "ProcessorParameter.h"
+#include "Transport.h"
+#include <AK/Noncopyable.h>
+#include <AK/StdLibExtras.h>
+#include <AK/Types.h>
+#include <LibCore/Object.h>
+
+namespace LibDSP {
+
+// A processor processes notes or audio into notes or audio. Processors are e.g. samplers, synthesizers, effects, arpeggiators etc.
+class Processor : public Core::Object {
+    C_OBJECT(Processor);
+    AK_MAKE_NONCOPYABLE(Processor);
+    AK_MAKE_NONMOVABLE(Processor);
+
+public:
+    virtual ~Processor()
+    {
+    }
+    Signal process(Signal const& input_signal)
+    {
+        VERIFY(input_signal.type == m_input_type);
+        auto processed = process_impl(input_signal);
+        VERIFY(processed.type == m_output_type);
+        return processed;
+    }
+    SignalType input_type() const { return m_input_type; }
+    SignalType output_type() const { return m_output_type; }
+    NonnullRefPtrVector<ProcessorParameter> parameters() const { return m_parameters; }
+
+private:
+    virtual Signal process_impl(Signal const& input_signal);
+    const SignalType m_input_type;
+    const SignalType m_output_type;
+
+protected:
+    Processor(NonnullRefPtr<Transport> transport, SignalType input_type, SignalType output_type)
+        : m_input_type(input_type)
+        , m_output_type(output_type)
+        , m_transport(transport)
```

Comment:
```suggestion
        , m_transport(move(transport))
```
This avoids an unnecessary additional `ref()` and `deref()` call.

Reply:
Why a pointer move?

Reply:
Because `transport` is a `NonnullRefPtr`, which isn't a primitive. In the copy constructor (`NonnullRefPtr(const NonnullRefPtr& other)`, `NonnullRefPtr.h:83`), there is a call to `other.add_ref()`, which calls `ref_if_not_null`, which calls `ptr->ref()` on the internal raw pointer. This increases the `m_ref_count` in the `RefCounted` class, which `Transport` inherits from. In the destructor of a `NonnullRefPtr`, there is a call to `unref()`, which decreases the `m_ref_count` and destroys the object if the reference count reaches zero.

What happens here is that the constructer is passed a `NonnullRefPtr` by value (so it is potentially copied). Then, `m_transport(transport)` copies that `NonnullRefPtr` again, which calls `ref()`. When the constructor returns, `transport`, so the `NonnullRefPtr`'s destructor is called, which calls `unref()`. If you instead `move()` the `NonnullRefPtr`, there are no calls to `ref()` and `unref()` since the count of `RefPtr`s pointing to the object remains constant.

The overhead that happens there isn't huge, but It's still good practice to do it properly.